ext/drbbdup/drbbdup.c

```diff
@@ -504,58 +504,76 @@ drbbdup_next_end(instr_t *instr)
  *  The returned instr list needs to be destroyed using instrlist_clear_and_destroy().
  */
 static instrlist_t *
-drbbdup_extract_single_bb_copy(void *drcontext, instrlist_t *bb, instr_t *start,
-                               OUT instr_t **next)
+drbbdup_extract_bb_copy(void *drcontext, instrlist_t *bb, instr_t *start,
+                        OUT instr_t **prev, OUT instr_t **post)
 {
     instrlist_t *case_bb = instrlist_create(drcontext);
 
     ASSERT(start != NULL, "start instruction cannot be NULL");
+    ASSERT(prev != NULL, "prev instr storage cannot be NULL");
+    ASSERT(post != NULL, "post instr storage cannot be NULL");
     ASSERT(instr_get_note(start) == (void *)DRBBDUP_LABEL_START,
            "start instruction should be a START label");
 
-    instr_t *instr = instr_get_next(start); /* Skip START label. */
-    while (instr != NULL && !drbbdup_is_at_end(instr)) {
-        instr_t *instr_cpy = instr_clone(drcontext, instr);
-        instrlist_append(case_bb, instr_cpy);
-        instr = instr_get_next(instr);
-    }
-    ASSERT(instr != NULL, "end instruction cannot be NULL");
-    ASSERT(!drbbdup_is_at_start(instr), "end cannot be at start");
-
-    /* Point to the next bb. */
-    if (next != NULL)
-        *next = drbbdup_next_start(instr);
+    *post = drbbdup_next_end(start);
+    ASSERT(*post != NULL, "end instruction cannot be NULL");
+    ASSERT(!drbbdup_is_at_start(*post), "end cannot be at start");
 
     /* Also include the last instruction in the bb if it is a
      * syscall/cti instr.
      */
     instr_t *last_instr = instrlist_last(bb);
     if (drbbdup_is_special_instr(last_instr)) {
         instr_t *instr_cpy = instr_clone(drcontext, last_instr);
-        instrlist_append(case_bb, instr_cpy);
+        instrlist_preinsert(bb, *post, instr_cpy);
     }
+    instrlist_cut(bb, *post);
+    *prev = start;
+    start = instr_get_next(start); /* Skip START label. */
+    instrlist_cut(bb, start);
+    instrlist_append(case_bb, start);
 
     return case_bb;
 }
 
+static void
+drbbdup_stitch_bb_copy(void *drcontext, instrlist_t *bb, instrlist_t *case_bb,
+                       instr_t *pre, instr_t *post)
+{
+    instr_t *last_instr = instrlist_last(case_bb);
+    if (drbbdup_is_special_instr(last_instr)) {
+        instrlist_remove(case_bb, last_instr);
+        instr_destroy(drcontext, last_instr);
+    }
+
+    instrlist_append(case_bb, post);
+    instr_t *instr = instrlist_first(case_bb);
+    instrlist_postinsert(bb, pre, instr);
+
+    instrlist_init(case_bb);
```

Comment:
Note: I could have used instrlist_append_instrlist() here but for some reason it is not exposed by dr core

Reply:
I guess i'll leave it how it is for now.